group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,448 @@
+     */
+    private MetadataImage metadataImage;
+
+    /**
+     * The group metadata manager.
+     */
+    private final GroupMetadataManager groupMetadataManager;
+
+    /**
+     * The maximum allowed metadata for any offset commit.
+     */
+    private final int offsetMetadataMaxSize;
+
+    /**
+     * The offsets keyed by topic-partition and group id.
+     */
+    private final TimelineHashMap<String, TimelineHashMap<TopicPartition, OffsetAndMetadata>> offsetsByGroup;
+
+    OffsetMetadataManager(
+        SnapshotRegistry snapshotRegistry,
+        LogContext logContext,
+        Time time,
+        MetadataImage metadataImage,
+        GroupMetadataManager groupMetadataManager,
+        int offsetMetadataMaxSize
+    ) {
+        this.snapshotRegistry = snapshotRegistry;
+        this.log = logContext.logger(OffsetMetadataManager.class);
+        this.time = time;
+        this.metadataImage = metadataImage;
+        this.groupMetadataManager = groupMetadataManager;
+        this.offsetMetadataMaxSize = offsetMetadataMaxSize;
+        this.offsetsByGroup = new TimelineHashMap<>(snapshotRegistry, 0);
+    }
+
+    /**
+     * Validates an OffsetCommit request.
+     *
+     * @param context The request context.
+     * @param request The actual request.
+     */
+    private void validateOffsetCommit(
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws ApiException {
+        Group group;
+        try {
+            // If the group does not exist and generation id is -1, the request comes from
+            // either the admin client or a consumer which does not use the group management
+            // facility. In this case, a so-called simple group is created and the request
+            // is accepted.
+            group = groupMetadataManager.getOrMaybeCreateSimpleGroup(
+                request.groupId(),
+                request.generationIdOrMemberEpoch() < 0
+            );
+        } catch (GroupIdNotFoundException ex) {
+            // Maintain backward compatibility. This is a bit weird in the
+            // context of the new protocol though.
+            throw Errors.ILLEGAL_GENERATION.exception();
+        }
+
+        // Validate the request based on the group type.
+        switch (group.type()) {
+            case GENERIC:
+                validateOffsetCommitForGenericGroup(
+                    (GenericGroup) group,
+                    request
+                );
+                break;
+
+            case CONSUMER:
+                validateOffsetCommitForConsumerGroup(
+                    (ConsumerGroup) group,
+                    context,
+                    request
+                );
+                break;
+        }
+    }
+
+    /**
+     * Validates an OffsetCommit request for a generic group.
+     *
+     * @param group     The generic group.
+     * @param request   The actual request.
+     */
+    public void validateOffsetCommitForGenericGroup(
```

Comment:
This validation logic is based on the code [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala#L1018).